components/brave_news/browser/brave_news_controller.cc

```diff
@@ -455,39 +459,86 @@ void BraveNewsController::GetImageData(const GURL& padded_image_url,
-    std::optional<std::vector<uint8_t>> args;
-    std::move(callback).Run(std::move(args));
+    std::move(callback).Run(std::nullopt);
     return;
   }
   const bool is_padded =
       (file_name.compare(file_name.length() - ending.length(), ending.length(),
                          ending) == 0);
-  VLOG(3) << "is padded: " << is_padded;
+  DVLOG(4) << "is padded: " << is_padded;
+  // Ensure we don't de-pad non PCDN images
+  if (is_padded && !padded_image_url.DomainIs(GetMatchingPCDNHostname())) {
+    DLOG(ERROR) << "Encountered a padded URL apparently not from PCDN: "
+                << padded_image_url.spec();
+    std::move(callback).Run(std::nullopt);
+    return;
+  }
   // Make the request
   private_cdn_request_helper_.DownloadToString(
       padded_image_url,
+      base::BindOnce(&BraveNewsController::OnGetImageDataFetched,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback),
+                     is_padded));
+}
+
+void BraveNewsController::OnGetImageDataFetched(GetImageDataCallback callback,
+                                                const bool is_padded,
+                                                const int response_code,
+                                                const std::string& body) {
+  // Handle the response
+  DVLOG(4) << "getimagedata response code: " << response_code;
+  // Attempt to remove byte padding if applicable
+  std::string_view body_payload(body.data(), body.size());
+  if (response_code != 200 ||
+      (is_padded &&
+       !brave::PrivateCdnHelper::GetInstance()->RemovePadding(&body_payload))) {
+    // Byte padding removal failed
+    std::move(callback).Run(std::nullopt);
+    return;
+  }
+  // Download (and optional unpadding) was successful
+  std::vector<uint8_t> image_bytes(body_payload.begin(), body_payload.end());
+#if BUILDFLAG(IS_ANDROID)
+  std::move(callback).Run(std::move(image_bytes));
+#else
+  if (is_padded) {
+    // Padded images are already re-encoded on the PCDN
+    std::move(callback).Run(std::move(image_bytes));
+    return;
+  }
+  // For security purposes, decode and re-encode the image,
+  // and don't just pass the bytes straight to WebUI.
+  data_decoder::DecodeImage(
+      &data_decoder_, image_bytes, data_decoder::mojom::ImageCodec::kDefault,
+      /*shrink_to_fit=*/true, data_decoder::kDefaultMaxSizeInBytes,
+      /*desired_image_frame_size=*/gfx::Size(),
+      base::BindOnce(&BraveNewsController::OnGetImageDataDecoded,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+#endif
+}
+
+void BraveNewsController::OnGetImageDataDecoded(GetImageDataCallback callback,
+                                                const SkBitmap& decoded_image) {
+  if (decoded_image.isNull()) {
+    DVLOG(1) << "Failed to decode image";
+    std::move(callback).Run(std::nullopt);
+    return;
+  }
+  // Encode on a thread and reply
+  base::ThreadPool::PostTaskAndReplyWithResult(
```

Comment:
I suggest to not use the default TaskTraits. Better to specific the task priority directly.

Reply:
SanitizedImageSource from Chromium uses default traits, but for this I now use:
```
{base::TaskPriority::USER_VISIBLE,
       base::TaskShutdownBehavior::CONTINUE_ON_SHUTDOWN}
```

Reply:
wow, `CONTINUE_ON_SHUTDOWN` is a dangerous one. Also all browser things (the global singletons, profiles, keyed services) could die when the task is running.
Do you really need it?

Reply:
Well, that's the opposite of what I intended to signal. Did I misinterpret this comment for `CONTINUE_ON_SHUTDOWN`?

```
  // Tasks posted with this mode which have not started executing before
  // shutdown is initiated will never run. Tasks with this mode running at
  // shutdown will be ignored (the worker will not be joined).
```

Compared to the comment for `SKIP_ON_SHUTDOWN` which seems to imply it will run and block shutdown
```
  // Tasks posted with this mode that have not started executing at
  // shutdown will never run. However, any task that has already begun
  // executing when shutdown is invoked will be allowed to continue and
  // will block shutdown until completion.
```

Perhaps this is why Chromium tends to not specify the traits and to use the defaults.

Reply:
discussed in DM. `CONTINUE_ON_SHUTDOWN` is a default ones.
`CONTINUE_ON_SHUTDOWN` also fits here.

Reply:
`CONTINUE_ON_SHUTDOWN` is in the branch